src/components/molecules/Visualizer/Plugin/PopupContainer/index.tsx

```diff
@@ -0,0 +1,68 @@
+import { computePosition, offset, autoUpdate } from "@floating-ui/dom";
+import {
+  forwardRef,
+  ForwardRefRenderFunction,
+  useRef,
+  useImperativeHandle,
+  useEffect,
+  RefObject,
+} from "react";
+
+import theme, { styled } from "@reearth/theme";
+
+import type { PopupPosition } from "../types";
+
+type AxesOffsets = {
+  mainAxis?: number;
+  crossAxis?: number;
+  alignmentAxis?: number | null;
+};
+
+export type PluginPopupInfo = {
+  id?: string;
+  position?: PopupPosition;
+  offset?: number | AxesOffsets;
+  ref?: RefObject<HTMLIFrameElement>;
+};
+
+type Props = {
+  shownPluginPopupInfo?: PluginPopupInfo;
+};
+
+const PopupContainer: ForwardRefRenderFunction<HTMLDivElement | undefined, Props> = (
+  { shownPluginPopupInfo },
+  ref,
+) => {
+  const innerRef = useRef<HTMLDivElement>(null);
+
+  useImperativeHandle(ref, () => innerRef.current as HTMLDivElement);
+
+  useEffect(() => {
+    if (!shownPluginPopupInfo?.ref?.current || !innerRef.current) return;
+    const cleanup = autoUpdate(shownPluginPopupInfo?.ref?.current, innerRef.current, () => {
+      if (!shownPluginPopupInfo?.ref?.current || !innerRef.current) return;
+      computePosition(shownPluginPopupInfo?.ref?.current, innerRef.current, {
+        placement: shownPluginPopupInfo.position ?? "bottom-start",
+        middleware: [offset(shownPluginPopupInfo.offset ?? 0)],
+      }).then(({ x, y }) => {
+        if (innerRef.current) {
+          innerRef.current.style.left = `${x}px`;
+          innerRef.current.style.top = `${y}px`;
+        }
+      });
+    });
```

Comment:
Since this is only being used in the useEffect in the return, and it is kinda long, can you just put this directly in the return? I feel like it would be easier to read.

Reply:
Updated.